packages/react-native-lockdown/src/index.js

```diff
@@ -0,0 +1,149 @@
+const { warn } = console
+const { assign } = Object
+
+const defaultGetRunModuleStatement = (moduleId) => `__r(${moduleId})`
+
+const validatePolyfills = (polyfills) => {
+  for (const polyfill of polyfills) {
+    if (typeof polyfill !== 'string') {
+      throw new Error(
+        `Expected polyfills to be an array of strings, but received ${typeof polyfill}`
+      )
+    }
+  }
+}
+
+module.exports = {
+  // simpler version that assumes react-native InitializeCore is the first thing that gets required.
+  lockdownSerializer: ({ hermesRuntime = false } = {}, originalConfig) => {
+    if (originalConfig.getRunModuleStatement) {
+      warn(
+        'LavaMoat: You are using getRunModuleStatement in serializer config. Lavamoat will attempt to wrap it without breaking it, but if you are doing something unusual, we might affect how it works'
+      )
+    }
+
+    const config = assign({}, originalConfig)
+
+    let firstCallGetRunModuleStatement = true
+    let warningTimer
+    const previousGetRunModuleStatement =
+      originalConfig.getRunModuleStatement ?? defaultGetRunModuleStatement
+
+    config.getRunModuleStatement = (moduleId) => {
+      const runModuleStatement = previousGetRunModuleStatement(moduleId)
+      if (firstCallGetRunModuleStatement) {
+        firstCallGetRunModuleStatement = false
+        warningTimer = setTimeout(() => {
+          warn(
+            'LavaMoat: getRunModuleStatement was only called once instead of at least twice, with react-native InitializeCore first and bundle entry second. This is suspicious. Lockdown might not be in effect.'
+          )
+        }, 50)
+        if (moduleId === 0) {
+          warn(
+            'LavaMoat: getRunModuleStatement was called with moduleId 0 first. That tends to represent the bundle entrypoint but react-native InitializeCore should run first. This is suspicious. Lockdown might not be in effect.'
+          )
+        }
+        return runModuleStatement + ';hardenIntrinsics();'
+      }
+      clearTimeout(warningTimer)
+      return runModuleStatement
+    }
+
+    if (!config.getPolyfills) {
+      config.getPolyfills = require('@react-native/js-polyfills')
+    } else if (typeof config.getPolyfills !== 'function') {
+      throw new Error('serializer.getPolyfills must be a function')
+    }
+    const ses = hermesRuntime
+      ? require.resolve('ses/hermes')
+      : require.resolve('ses')
+    const repair = require.resolve('./repair.js')
+    const harden = require.resolve('./harden.js')
+
+    config.getPolyfills = () => {
+      // const polyfills = config.getPolyfills() // RangeError: Maximum call stack size exceeded
+      const polyfills = require('@react-native/js-polyfills')() // TODO: respect originalConfig.getPolyfills
+      validatePolyfills(polyfills)
+      return [ses, repair, ...polyfills, harden]
```

Comment:
<img width="465" alt="Screenshot 2025-04-03 at 4 47 03 pm" src="https://github.com/user-attachments/assets/980e2e6c-f7f6-4226-8a42-730ca228c6f5" />

```suggestion
      return [ses, repair, ...polyfills]
```

this is because we harden twice 1 getPolyfills 2 getRunModuleStatement
(under-the-hood `hardenIntrinsics = repairIntrinsics(options)`)

1 getPolyfills, too early

<img width="1033" alt="Screenshot 2025-04-03 at 4 53 41 pm" src="https://github.com/user-attachments/assets/0cbff4c5-7668-42c1-9e54-65514fab78af" />

2 getRunModuleStatement

<img width="1032" alt="Screenshot 2025-04-03 at 4 57 21 pm" src="https://github.com/user-attachments/assets/aedfc587-2db7-4221-a66a-8dd9a14acecc" />

Reply:
create endo issue, SES_ALREADY_LOCKED_DOWN error was conceived before repair/harden

Reply:
- https://github.com/endojs/endo/issues/2760